Pointer arithmetics don't work in forward mode

This is similar to #972, only now it's about pointer arithmetic that backward mode supports but not the forward mode.

It would be nice if backward mode and forward mode have the same level of C++ language support to avoid more surprises like this :slightly_smiling_face:

Reproducer:
```c++
// Compile with clang++ -std=c++11 -fplugin=install/lib/clad.so -I./install/include cladIssue.cxx -o cladIssue

#include "clad/Differentiator/Differentiator.h"

#include <iostream>
#include <vector>

double wrapper(double* params, double const* constants) {
   double c0 = *(constants + 0);
   return params[0] * c0;
}

int main()
{
   auto diff = clad::differentiate(wrapper, "params[0]"); // doesn't work
   auto grad = clad::gradient(wrapper, "params"); // works!

   double x = 2.0;
   double y = 2.0;
   double result = 0.0;
   grad.execute(&x, &y, &result);

   std::cout << result << std::endl;
   std::cout << diff.execute(&x, &y) << std::endl;
}
```
Compiler output:
```txt
cladIssue.cxx:9:30: warning: attempt to differentiate unsupported operator,  derivative                          set to 0
   double c0 = *(constants + 0);
                             ^
error: indirection requires pointer operand ('int' invalid)
1 warning and 1 error generated.
```

--- a/lib/Differentiator/BaseForwardModeVisitor.cpp
+++ b/lib/Differentiator/BaseForwardModeVisitor.cpp
@@ -1460,13 +1460,13 @@ BaseForwardModeVisitor::VisitBinaryOperator(const BinaryOperator* BinOp) {
     // including it as a Stmt_dx. Moreover, the fact that Stmt_dx is left
     // nullptr is used for treating expressions like ((A && B) && C) correctly.
     return StmtDiff(opDiff, nullptr);
-  }
-  if (!opDiff) {
+  } else {
     // FIXME: add support for other binary operators
     unsupportedOpWarn(BinOp->getEndLoc());
     opDiff = ConstantFolder::synthesizeLiteral(m_Context.IntTy, m_Context, 0);
   }
-  opDiff = folder.fold(opDiff);
+  if (opDiff)
+    opDiff = folder.fold(opDiff);
   // Recover the original operation from the Ldiff and Rdiff instead of
   // cloning the tree.
   Expr* op;
@@ -1499,11 +1499,19 @@ BaseForwardModeVisitor::DifferentiateVarDecl(const VarDecl* VD,
   // This may not necessarily be true in the future.
   VarDecl* VDClone =
       BuildVarDecl(VD->getType(), VD->getNameAsString(), initDiff.getExpr(),
-                   VD->isDirectInit(), nullptr, VD->getInitStyle());
+                   VD->isDirectInit(), /*TSI=*/nullptr, VD->getInitStyle());
   // FIXME: Create unique identifier for derivative.
-  VarDecl* VDDerived = BuildVarDecl(
-      VD->getType(), "_d_" + VD->getNameAsString(), initDiff.getExpr_dx(),
-      VD->isDirectInit(), nullptr, VD->getInitStyle());
+  Expr* initDx = initDiff.getExpr_dx();
+  if (VD->getType()->isPointerType() && !initDx) {
+    // initialize with nullptr.
+    // NOLINTBEGIN(cppcoreguidelines-owned-memory)
+    initDx =
+        new (m_Context) CXXNullPtrLiteralExpr(VD->getType(), VD->getBeginLoc());
+    // NOLINTEND(cppcoreguidelines-owned-memory)
+  }
+  VarDecl* VDDerived =
+      BuildVarDecl(VD->getType(), "_d_" + VD->getNameAsString(), initDx,
+                   VD->isDirectInit(), /*TSI=*/nullptr, VD->getInitStyle());
   m_Variables.emplace(VDClone, BuildDeclRef(VDDerived));
   return DeclDiff<VarDecl>(VDClone, VDDerived);
 }

--- a/test/ForwardMode/MemberFunctions.C
+++ b/test/ForwardMode/MemberFunctions.C
@@ -719,7 +719,7 @@ public:
   // CHECK-NEXT:       double _d_j = 0;
   // CHECK-NEXT:       SimpleFunctions _d_this_obj;
   // CHECK-NEXT:       SimpleFunctions *_d_this = &_d_this_obj;
-  // CHECK-NEXT:       double *_d_p;
+  // CHECK-NEXT:       double *_d_p = nullptr;
   // CHECK-NEXT:       double *p;
   // CHECK-NEXT:       _d_p = _d_this->arr[1];
   // CHECK-NEXT:       p = this->arr[1];

--- a/test/ForwardMode/Pointer.C
+++ b/test/ForwardMode/Pointer.C
@@ -203,6 +203,16 @@ double fn9(double* params, const double *constants) {
 // CHECK-NEXT:     return 1. * c0 + params[0] * _d_c0;
 // CHECK-NEXT: }
 
+double fn10(double *params, const double *constants) {
+  double c0 = *(constants + 0);
+  return params[0] * c0;
+}
+
+// CHECK: double fn10_darg0_0(double *params, const double *constants) {
+// CHECK-NEXT:     double _d_c0 = 0;
+// CHECK-NEXT:     double c0 = *(constants + 0);
+// CHECK-NEXT:     return 1. * c0 + params[0] * _d_c0;
+// CHECK-NEXT: }
 
 int main() {
   INIT_DIFFERENTIATE(fn1, "i");
@@ -229,6 +239,10 @@ int main() {
   double constants[] = {5.0};
   auto fn9_dx = clad::differentiate(fn9, "params[0]");
   d_param = fn9_dx.execute(params, constants);
+  printf("{%.2f}\n", d_param); // CHECK-EXEC: {5.00}
+
+  auto fn10_dx = clad::differentiate(fn10, "params[0]");
+  d_param = fn10_dx.execute(params, constants);
   printf("{%.2f}\n", d_param); // CHECK-EXEC: {5.00}
 }
 

